server/src/main/java/io/spine/server/bus/ValidatingFilter.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2017, TeamDev Ltd. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.bus;
+
+import com.google.common.base.Optional;
+import com.google.protobuf.Message;
+import io.spine.base.Error;
+import io.spine.core.IsSent;
+import io.spine.core.MessageEnvelope;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.spine.server.bus.Buses.reject;
+
+/**
+ * A filter validating the {@linkplain MessageEnvelope envelopes} with the given
+ * {@link EnvelopeValidator}.
+ *
+ * @author Dmytro Dashenkov
+ */
+final class ValidatingFilter<E extends MessageEnvelope<T>, T extends Message>
+        extends AbstractBusFilter<E> {
+
+    private final Bus<T, E, ?, ?> bus;
+    private final EnvelopeValidator<E> validator;
+
+    ValidatingFilter(Bus<T, E, ?, ?> bus) {
+        super();
+        this.bus = bus;
```

Comment:
Let's have an internal `IdConverter` in `Bus` and store the reference to it, not to the whole `Bus`.